---
title: "GRCMPB Use_Case"
author: "Brandon Ngwa"
date: "2024-10-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load the libraries
```{r}

library(tidyverse)
library(devtools)
library(ggrepel)
library(ggthemes)
library(sf)
library(readxl)

```

```{r}
#lint("C:/Users/bngwa/Documents/Brandon/MPB_grcMalaria/R/Mutation_Frequency_Plots.R", linters = default_linters)
```


## Create your final GRC Data 
```{r }

GRC_Data <-
    combine_grc_sheets(input_folder="C:/Users/bngwa/Documents/Brandon/GDA_Markdown/All_GRC_Reads_Gambia",
                       country = "Gambia", 
                       save_output = TRUE)
```


## Create the drug status columns
```{r, echo=FALSE}

GRC_Data <-
  gene_classifier(df = GRC_Data, 
                  drug_column = "Chloroquine",
                  save_output = TRUE)
```


## Create your meta data for the maps
```{r}
## load the shapefile for Gambia first
GMB <- st_read("C:/Users/bngwa/Documents/Brandon/GDA_Markdown/geoBoundaries-GMB-ADM3-all/geoBoundaries-GMB-ADM3-all/geoBoundaries-GMB-ADM3_simplified.shp")

## Read the logitude and latitude data
LongLat <- read_excel("C:/Users/bngwa/Documents/Brandon/GDA_Markdown/LongLat_data.xlsx")

geo_data <- 
mapping_data(shapefile = GMB ,
            long_lat_data = LongLat,
            location_col = "Location",
            long_col = "long",
            lat_col = "lat" )
```


## Create Sample Count Map
```{r,echo=FALSE, out.width="100%" }

## You can creat a single or range of years to filter buy before generating the plots
Periods <-
  list(list(name="2021", type ="year", start="2021"),
       list(name="2017-19",  type= "period", start="2017",  end="2019"))

sample_count_map(df = GRC_Data, 
                 map_data = geo_data,
                 time = NULL,
                 breaks = c(10, 100, 200, 300),
                 label_size = 2.5, 
                 scale_circle_size = 11,
                 save_output = TRUE)

```


## Create Distrubution table and barchart 
```{r, echo=FALSE, out.width="100%}

drug_distribution(df = GRC_Data, 
                  drug_col = "Chloroquine",
                  save_output = TRUE,
                  time = NULL,
                  colors = c("resistant" = "#525CEB",
                             "mixed_resistant" = "#808000",
                             "sensitive" = "#800000"))
```


## Drug Resistance Prevalence Proportion Maps
```{r, echo=FALSE, out.width="100%}

drug_distribution_pm(df = GRC_Data, 
                     drug_col = "Chloroquine",
                     save_output = TRUE,
                     time = NULL,
                     map_data = geo_data,
                     label_size = 2.5,
                     circle_num_size = 3.1, 
                     scale_circle_size = 10)
               
```


## Mutation Frequency table and plots 
```{r, echo=FALSE, out.width="100%}

Mutation_Frequency(df = GRC_Data, 
                   gene = "pfcrt", 
                   gene_col = "PfCRT", 
                   drug_col = "Chloroquine",
                   save_output = TRUE,
                   time = NULL,
                   mData = geo_data,
                   label_size = 2.5,
                   circle_num_size = 3.1, 
                   scale_circle_size = 10,
                   include_mixed = FALSE)

```


## Haplotype Proportion plots
```{r, echo=FALSE, out.width="100%}

haplotype_proportion(df = GRC_Data, 
                     gene_col = "PfCRT", 
                     drug_col = "Chloroquine",
                     save_output = TRUE,
                     time = NULL,
                     map_data = geo_data,
                     label_size = 2.5,
                     sacle_piechart_size = 0.035)

```


## Filter SNPs and Smaples for missigness
```{r}
barcode_data <- 
filter_snp_x_samples(df = GRC_Data, 
                     m_threshold = 0.40)
head(barcode_data)
```


## Diversity Map
```{r, echo=FALSE, out.width="100%}

diversity_map(df = GRC_Data, 
              drug_col = "Chloroquine",
              snp_data = barcode_data, 
              map_data = geo_data,
              label_size = 2.5,
              circle_num_size = 3.1, 
              scale_circle_size = 10,
              time = NULL,
              save_output = TRUE)
```


## Calculate the Identity By State (IBS) and save it as a dataframe.
```{r, echo=FALSE, out.width="100%}
IBS_Data <- 
generate_ibs_data(df = GRC_Data, 
                  snp_data = barcode_data,
                  drug_col = "Chloroquine")

IBS_Data$IBS_Histogram
```


## Summarise IBS by the same location and draw the proportion map
```{r}

ibs_data_sl(melted_ibs_matrix = IBS_Data$IBS_Melted_Matrix,
            ibs_th = 0.75,
            map_data = geo_data,
            label_size = 2.5,
            circle_num_size = 3.1, 
            scale_circle_size = 10,
            save_output = TRUE)
```


## Summarise IBS by diffrent location and draw the propoertion map
```{r}

ibs_data_dl(melted_ibs_matrix = IBS_Data$IBS_Melted_Matrix,
            ibs_threshold = 0.75,
            map_data = geo_data,
            label_size = 5,
            breaks = seq(1,9,2),
            curve_degree = 0.5,
            save_output = TRUE)

```


## Generate IBS_heatmap
```{r}
 
ibs_heat_map(df = GRC_Data,
             snp_data = barcode_data,
             ibs_matrix = IBS_Data$IBS_matrix,
             map_data = geo_data,
             save_output = TRUE,
             drug_col = "Chloroquine")

```


## Generate the PCoA Plots
```{r}

pcoa_plots(ibs_matrix = IBS_Data$IBS_matrix,
           df = GRC_Data,
           circle_size = 4, 
           save_output = TRUE,
           drug_col = "Chloroquine")

```


```{r}
nj_tree(ibs_matrix = IBS_Data$IBS_matrix,
        df = GRC_Data,
        tippoint_size = 4,
        line_size = 0.6,
        save_output = TRUE,
        drug_col = "Chloroquine")

```


